wgsl/index.bs

```diff
@@ -3959,6 +3959,278 @@ TODO: deduplicate these tables
 
 ## Texture built-in functions ## {#texture-builtin-functions}
 
+### Sample ### {#texture-sample-functions}
+
+Sampling from a texture uses the `textureSample` function:
+
+```
+<result> textureSample(<texture>, <sampler>, <coordinates> [, <named-arguments>])
+```
+
+Where:
+
+* `<result>` is the texture's parameterized type when using [sampled](#sampled-texture-type)
+    or [read-only](#texture-ro) texture formats.<br>
+* `<result>` is `f32` when using [depth](#texture-depth) texture types.
+
+* `<texture>` is one of the [sampled](#sampled-texture-type),
+    [read-only](#texture-ro) or [depth](#texture-depth) texture types.
+
+* `<sampler>` is a [sampler type](#sampler-type).
+
+* `<coordinates>` are the texture coordinates used for sampling.
+    The type for the texture coordinates depends on the `<texture>` type and is
+    either a `f32`, `vec2<f32>`, `vec3<f32>` or `vec4<f32>`:
+
+    <table class='data'>
+      <thead>
+        <tr><td>`<texture>` type<td>`<coordinates>` type
+      </thead>
+      <tr><td> `texture_sampled_1d<type>`         <td> `f32`
+      <tr><td> `texture_sampled_1d_array<type>`   <td> `vec2<f32>`
+      <tr><td> `texture_sampled_2d<type>`         <td> `vec2<f32>`
+      <tr><td> `texture_sampled_2d_array<type>`   <td> `vec3<f32>`
+      <tr><td> `texture_sampled_3d<type>`         <td> `vec3<f32>`
+      <tr><td> `texture_sampled_cube<type>`       <td> `vec3<f32>`
+      <tr><td> `texture_sampled_cube_array<type>` <td> `vec4<f32>`
+      <tr><td> `texture_ro_1d<type>`              <td> `f32`
+      <tr><td> `texture_ro_1d_array<type>`        <td> `vec2<f32>`
+      <tr><td> `texture_ro_2d<type>`              <td> `vec2<f32>`
+      <tr><td> `texture_ro_2d_array<type>`        <td> `vec3<f32>`
+      <tr><td> `texture_ro_3d<type>`              <td> `vec3<f32>`
+      <tr><td> `texture_depth_2d`                 <td> `vec2<f32>`
+      <tr><td> `texture_depth_2d_array`           <td> `vec3<f32>`
+      <tr><td> `texture_depth_cube`               <td> `vec3<f32>`
+      <tr><td> `texture_depth_cube_array`         <td> `vec4<f32>`
+    </table>
+
+    Array textures (with `_array` suffix) use the last component of `<coordinates>`
+    to specify array index.
+
+
+<div class='example' heading="Sample">
+  <xmp>
+    color = textureSample(tex2D, s, coords);
+  </xmp>
+</div>
+
+Following the `<coordinates>` parameter, a number of additional named arguments may
+be passed for the sample operation:
+
+#### Sample with texel offset #### {#texture-sample-with-offset}
+
+```
+<result> textureSample(<texture>, <sampler>, <coordinates>, offset: <offset> [, <named-arguments>])
+```
+
+Applies the integer texel offset `<offset>` to the texture coordinate before sampling.
```

Comment:
we also need to say that this is applied to the unnormalized texture coordinate (as opposed to the normalized one that the user provided). We also need to double check if the addressing modes of the sampler (e.g. wrapping) are applied before this offset or after.

Reply:
> we also need to say that this is applied to the unnormalized texture coordinate (as opposed to the normalized one that the user provided).

Done.

> We also need to double check if the addressing modes of the sampler (e.g. wrapping) are applied before this offset or after.

Ack. Added a TODO for now.

Reply:
> We also need to double check if the addressing modes of the sampler (e.g. wrapping) are applied before this offset or after.

The order appears to be well defined for HLSL and Vulkan - these both apply offsets before texture wrapping:

[HLSL - Calculating texel positions](https://docs.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-to-sample#calculating-texel-positions).

[Vulkan - 15.5.8. (s,t,r,q,a) to (u,v,w,a) Transformation](https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/chap16.html#textures-normalized-to-unnormalized).

MSL is more vague (6.10 Texture Functions):
> The offset is an integer value applied to the texture coordinate before looking up each pixel.

I'm not sure if there's some better clarification hidden away somewhere else.

I'll gone with the pre-wrap definition for now. In the unlikely situation we find MSL behaves differently, we might need to emit some emulation logic.